docs/contributing/3.0-development/app-layout.rst

```diff
@@ -0,0 +1,241 @@
+
+* No models are defined in the ``__init__.py`` of ``pulp.app.models``.
+* All models are defined in submodules located in the ``pulp.app.models`` module
+  directory (where its ``__init__.py`` can be found).
+* The `__init__.py`` in ``pulp.app.models`` should consist only of import statements,
+  ordered to prevent any circular import issues that may result based on the imports
+  that are done in any included submodules.
+* Any models defined in submodules in ``pulp.app.models`` namespace must be imported
+  from the ``pulp.app.models`` namespace, not the submodule in which they are defined.
+  Yes: ``from pulp.app.models import PulpModel``,
+  No: ``from pulp.app.models.pulp import PulpModel``.
+* When adding new models, they must be imported into the ``pulp.app.models``
+  ``__init__.py``, so that they are available to be imported by any other Pulp Platform
+  components that use them from the ``pulp.app.models`` namespace.
+* Imports done inside any submodules should be relative, e.g.
+  ``from .submodule import identifier`` or ``from . import submodule``, avoiding the
+  creation of circular imports.
+* Imports done inside the module's ``__init__.py`` should be relative and explict, e.g.
+  ``from .submodule import identifier1, identifier2``, NOT
+  ``from submodule import identifier1, identifier2`` and NOT
+  ``from .submodule import *``.
+
+Any module in ``pulp.app`` may be broken up in this way, such as
+:mod:`pulp.app.serializers` or :mod:`pulp.app.viewsets`, should do so in such a way
+that renders the implementation invisible to anyone importing from that module.
+
+pulp.app
+--------
+
+pulp.app is the package containing the core Pulp Platform Django application.
+This package contains all of the Pulp Platform models, serializers, and
+viewsets required to assemble Pulp's REST API and underlying database.
+
+pulp.app.apps
+^^^^^^^^^^^^^
+
+This module defines the base class used by all Pulp plugins to identify them to
+the Pulp Platform as plugins to be imported by Pulp, as well as the Pulp Platform
+application config class itself, which is exposed to plugin writers via the Plugin API.
+
+pulp.app.settings
+^^^^^^^^^^^^^^^^^
+
+This is the main settings module for the platform Django project, which puts together
+all of the various Django applications that the Pulp Platform depends on to function,
+as well as the Pulp Platform application itself and its plugins.
+
+Many things are defined in here, including the database settings, logging configuration,
+REST API settings, etc. This file also finds and registers Pulp plugins with the Pulp
+Platform Django Project, using the ``pulp.plugin`` entry point.
+
+In order to use django-related tools with the Pulp Platform, the platform must be installed,
+and the ``DJANGO_SETTINGS_MODULE`` environment var must be set to
+:mod:`pulp.app.settings`.
+
+pulp.app.urls
+^^^^^^^^^^^^^
+
+This module lists all API endpoints exposed by the Pulp REST API, including all viewsets
```

Comment:
Depends on your perspective. Saying it defines the endpoints isn't much clearer, so I'll think about how to word this better.